queue/index.js

```diff
@@ -0,0 +1,47 @@
+export class Queue {
+  constructor(app, actionTypes, key) {
+    this.app = app
+    this.actionTypes = actionTypes
+    this.key = key
+    this.data = []
+    this.processing = false
+  }
+
+  add(action, meta) {
+    this.data.push(action)
+    this.data.push(meta)
+    if (!this.processing) {
+      this.processing = true
+      this.processAction(action, meta)
+    }
+  }
+
+  remove() {
+    this.data = this.data.slice(2)
+    if (this.data.length) {
+      this.processAction(this.data[0], this.data[1])
+    } else {
+      this.processing = false
+    }
+  }
+
+  async processAction(action, meta) {
+    let client = this.app.connected.get(this.key)
```

Comment:
What if action will be processed when client will be already disconnected

1. Client A connects
2. Client B connects
3. Client A adds action 1
4. Queue is busy with action 1 processing
5. Client B adds action 2
6. Client B disconnects
7. Queue finished action 1 processing
8. Queue go to action 2, but Client B is already disconnected

Reply:
This should not cause any problems. As far as I understand, the `connected` Map in `BaseServer` does not get destroyed until the server itself is turned off. So this is what should happen in this scenario:
After step 8, the action is probably already marked as `synced` (or, otherwise not in queue at all) as everything up to `ServerClient.filter` happens in parallel, not in queue, the client will not be re-sending this action again if reconnected.
Similarly to the `connected` map in `BaseServer`, the `queues` map is not destroyed until server shuts down, so the queued action will be added to the log even if the client is already disconnected.

Reply:
Maybe the `client` variable here should be renamed to `serverClient` for clarity?

Reply:
> As far as I understand, the connected Map in BaseServer does not get destroyed until the server itself is turned off

The problem is that `ServerClient` remove itself from `server.connected` on client’s disconnect

https://github.com/logux/server/blob/main/server-client/index.js#L99

Reply:
I see now. Maybe I can add a check to a `destroy` function, so that it would wait for queue to complete, or maybe it would be even better to rewrite `filter` function in `queue.processAction` (from this line: https://github.com/logux/server/blob/main/server-client/index.js#L207).
Also, I should add a way to delete a queue if a respective client is disconnected and all actions are processed.

Reply:
>  I can add a check to a destroy function, so that it would wait for queue to complete

`server.connected` is used to detect current active clients. We can’t change behavior.

We should be ready to process any action without a `ServerClient` instance. What if server will add its own action by `server.log.add(action, meta)`?

Reply:
Ok, I will add ability to process action without `ServerClient`, thank you for the pointers

Reply:
Done